lark/parsers/lalr_parser.py

```diff
@@ -35,84 +35,116 @@ def parse(self, *args):
-                del state_stack[-size:]
-                del value_stack[-size:]
+                raise UnexpectedToken(token, expected, state=state, puppet=None)
+
+            assert arg != self.end_state
+
+            if action is Shift:
+                # shift once and return
+                assert not is_end
+                state_stack.append(arg)
+                value_stack.append(token)
+                return arg
             else:
-                s = []
+                # reduce+shift as many times as necessary
+                rule = arg
+                size = len(rule.expansion)
+                if size:
+                    s = value_stack[-size:]
+                    del state_stack[-size:]
+                    del value_stack[-size:]
+                else:
+                    s = []
+
+                value = self.callbacks[rule](s)
+
+                _action, new_state = states[state_stack[-1]][rule.origin.name]
+                assert _action is Shift
+                state_stack.append(new_state)
+                value_stack.append(value)
+
+                if is_end and state_stack[-1] == self.end_state:
+                    return value_stack[-1]
 
-            value = self.callbacks[rule](s)
+class _Parser:
+    def __init__(self, parse_table, callbacks, debug=False):
+        self.parse_table = parse_table
+        self.callbacks = callbacks
+        self.debug = debug
 
-            _action, new_state = states[state_stack[-1]][rule.origin.name]
-            assert _action is Shift
-            state_stack.append(new_state)
-            value_stack.append(value)
+    def parse(self, lexer, start, value_stack=None, state_stack=None):
+        parser_state = ParserState(self.parse_table, lexer, self.callbacks, start, state_stack, value_stack)
+        return self.parse_from_state(parser_state)
 
+    def parse_from_state(self, state):
         # Main LALR-parser loop
         try:
-            for token in stream:
-                while True:
-                    action, arg = get_action(token)
-                    assert arg != end_state
-
-                    if action is Shift:
-                        state_stack.append(arg)
-                        value_stack.append(token)
-                        if set_state: set_state(arg)
-                        break # next token
-                    else:
-                        reduce(arg)
+            token = None
+            for token in state.lexer.lex(state):
+                state.feed_token(token)
+
+            token = Token.new_borrow_pos('$END', '', token) if token else Token('$END', '', 0, 1, 1)
+            return state.feed_token(token, True)
+        except UnexpectedInput as e:
+            try:
+                e.puppet = ParserPuppet(self, state, state.lexer)
+            except NameError:
+                pass
+            if isinstance(e, UnexpectedCharacters):
+                s = state.lexer.state
+                p = s.line_ctr.char_pos
```

Comment:
This makes an assumption over the layout of the LexerState (which is something I though we said we don't do?)

And I am also not sure what you are trying to accomplish here anyway.

Reply:
`UnexpectedCharacters` is raised when the lexer cannot match any token. That means it won't advance, and resuming a parse will just cause it to return to the same character and fail with the same error.

This advances the character pos by 1, so it will skip the "bad" character if the parse is resumed.

Reply:
> This makes an assumption over the layout of the LexerState (which is something I though we said we don't do?)

Why not? It's a defined class in `lexer.py`

I don't remember the context. Why shouldn't we make any assumption about it?

Reply:
This is something that should done in `on_error` or in `Lark.parse`, but not here, if at all. This would as example also break when you try to find the character that was blocking via `text[line_ctr.char_pos]`. I see that the goal is to make `lambda e: True` work, but I think a better solution would be to export maybe a `skip_error` function that does all this.

Once again, this might break with CustomLexer, and even if we don'T expose the new Lexer protocol now, it will break in 1.0.

Reply:
gitter chat Sep 26. (Can you somehow to a perma link?). I am using a custom lexer with a custom lexerstate for Lark-language server to allow `on_error` to skip over unknown tokens. Check my branch over there. You are here assuming that the lexer state has a `line_ctr` attribute which is not true. (and I could provided it, but it is completly useless since the tokens are preparsed.)

Reply:
But then your lexer would never throw `UnexpectedCharacters`, right?

Reply:
That is true. But if we required that (a Lexer that might throw `UnexpecedCharacters` has to have a `LexerState` with a `line_ctr` attribute we have to clearly document that somewhere) And I can easily imagine situation where position information might be unusable (A lexer that directly reads from a stream, or ask the user for new input if the internal buffer is empty like a shell). Also note that this would create an AttributeException even if we don't use `on_error`.

Reply:
You're right, it wasn't in the right place.

I refactored it into `Lark.parse()`. Now it won't trigger if `on_error` is None.

Reply:
I am still not completely certain that we shouldn't let the user explicit optin/optout of this behavior, but we can't wait till this because a problem. Since we aren't exposing the interface anyway, we can wait till 1.0 to make a final decision.